recipe-client-addon/test/unit/test_NormandyApi.js

```diff
@@ -126,6 +142,73 @@ add_task(withMockApiServer(async function test_fetchRecipes() {
   equal(recipes[0].name, "system-addon-test");
 }));
 
+add_task(async function test_fetchRecipes_canonical_mismatch() {
+  const getApiUrl = sinon.stub(NormandyApi, "getApiUrl").resolves("http://localhost/recipes/");
+
+  // Recipe is non-canonical (it has whitespace, properties are out of order)
+  const response = new MockResponse(`[
+    {
+      "recipe": {"b": 1, "a": 2},
+      "signature": {"signature": "", "x5u": ""}
+    }
+  ]`);
+  const get = sinon.stub(NormandyApi, "get").resolves(response);
+
+  try {
+    await NormandyApi.fetchRecipes();
+    ok(false, "fetchRecipes did not throw for canonical JSON mismatch");
+  } catch (err) {
+    ok(err instanceof InvalidSignatureError, "Error was not an InvalidSignatureError");
+    ok(/Canonical/.test(err), "Error was not due to canonical JSON mismatch");
+  }
+
+  getApiUrl.restore();
+  get.restore();
+});
+
+add_task(async function test_fetchRecipes_validation_error() {
+  const getApiUrl = sinon.stub(NormandyApi, "getApiUrl").resolves("http://localhost/recipes/");
+
+  // Mock two URLs: recipes and the x5u
+  const get = sinon.stub(NormandyApi, "get").callsFake(async url => {
+    if (url.endsWith("recipes/")) {
+      return new MockResponse(CanonicalJSON.stringify([
+        {
+          recipe: {a: 1, b: 2},
+          signature: {signature: "invalidsignature", x5u: "http://localhost/x5u/"},
+        },
+      ]));
+    } else if (url.endsWith("x5u/")) {
+      return new MockResponse("certchain");
+    }
+
+    return null;
+  });
+
+  // Validation should fail due to a malformed x5u and signature.
+  try {
+    await NormandyApi.fetchRecipes();
+    ok(false, "fetchRecipes did not throw for a validation error");
+  } catch (err) {
+    ok(err instanceof InvalidSignatureError, "Error was not an InvalidSignatureError");
+    ok(/signature/.test(err), "Error was not due to a validation error");
+  }
+
+  getApiUrl.restore();
+  get.restore();
+});
+
+const invalidSignatureServer = makeMockApiServer(do_get_file("invalid_recipe_signature_api"));
```

Comment:
Signature validation can fail for two reasons:

- If the data passed in is malformed (like if the x5u cert chain isn't in pem format) it throws an exception. This is what the first test handles.
- If the data is in the correct format but the signature doesn't match the content, it returns false. This is what the second test handles.

Uptake has a single status for signature issues, and I used it for both of these without thinking about it too much; on one hand, we could mark malformed data problems as a server issue, but on the other hand, it's still a problem that was raised during signature verification.

One is done using mocking and the other isn't mostly because I wrote the mock one first, and when I wanted to write the second one, the test data was long enough that I didn't want to keep it in JS. If we care, I could have them both use a mock API server for consistency's sake.